amazon-kinesis-client/src/main/java/software/amazon/kinesis/coordinator/Scheduler.java

```diff
@@ -423,63 +436,173 @@ void runProcessLoop() {
         final Set<StreamIdentifier> streamsSynced = new HashSet<>();
 
         if (shouldSyncStreamsNow()) {
-            final Map<StreamIdentifier, StreamConfig> newStreamConfigMap = new HashMap<>();
-            // Making an immutable copy
-            newStreamConfigMap.putAll(multiStreamTracker.streamConfigList().stream()
-                    .collect(Collectors.toMap(sc -> sc.streamIdentifier(), sc -> sc)));
-
-            // This is done to ensure that we clean up the stale streams lingering in the lease table.
-            syncStreamsFromLeaseTableOnAppInit();
-
-            for (StreamIdentifier streamIdentifier : newStreamConfigMap.keySet()) {
-                if (!currentStreamConfigMap.containsKey(streamIdentifier)) {
-                    log.info("Found new stream to process: " + streamIdentifier + ". Syncing shards of that stream.");
-                    ShardSyncTaskManager shardSyncTaskManager = createOrGetShardSyncTaskManager(newStreamConfigMap.get(streamIdentifier));
-                    shardSyncTaskManager.syncShardAndLeaseInfo();
-                    currentStreamConfigMap.put(streamIdentifier, newStreamConfigMap.get(streamIdentifier));
-                    streamsSynced.add(streamIdentifier);
-                } else {
-                    if (log.isDebugEnabled()) {
-                        log.debug(streamIdentifier + " is already being processed - skipping shard sync.");
+            final MetricsScope metricsScope = MetricsUtil.createMetricsWithOperation(metricsFactory, MULTI_STREAM_TRACKER);
+
+            try {
+
+                final Map<StreamIdentifier, StreamConfig> newStreamConfigMap = new HashMap<>();
+                final Duration waitPeriodToDeleteOldStreams = formerStreamsLeasesDeletionStrategy.waitPeriodToDeleteFormerStreams();
+                // Making an immutable copy
+                newStreamConfigMap.putAll(multiStreamTracker.streamConfigList().stream()
+                        .collect(Collectors.toMap(sc -> sc.streamIdentifier(), sc -> sc)));
+
+                List<MultiStreamLease> leases;
+
+                // This is done to ensure that we clean up the stale streams lingering in the lease table.
+                if (!leasesSyncedOnAppInit && isMultiStreamMode) {
+                    leases = fetchMultiStreamLeases();
+                    syncStreamsFromLeaseTableOnAppInit(leases);
+                    leasesSyncedOnAppInit = true;
+                }
+
+                // For new streams discovered, do a shard sync and update the currentStreamConfigMap
+                for (StreamIdentifier streamIdentifier : newStreamConfigMap.keySet()) {
+                    if (!currentStreamConfigMap.containsKey(streamIdentifier)) {
```

Comment:
Is there any leaks happening for currentStreamConfigMap and streamDeletionMap. Does the entries get removed from the list or never ?